text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+The approach described here is to provide a messaging system whereby modules publish relevant data when ready, for other modules to consume and do their part. Modules which participate in a particular workflow can subscribe to the relevant messages. The advantage of this particular solution is that it fits well with the microservices architecture and preserves the independence of the modules. Each module plays its part but remains unaware of any role played by other modules. It does not require a centralized workflow controller.
+
+## Detailed Explanation/Design
+
+### Requirements
+#### Functional Requirements
+- A business module (back-end) can define a set of events it can send (publish). It can also declare and receive (subscribe to) event of different types.
+
+- An “Event Descriptor” must be provided to define new event types.
+  1. The event definition produces a unique “Event Type”
+  2. The “Event Descriptor” must have a detailed description for the event type and payloads which can be received within events of this type.
+  3. “Event Descriptor” does not support versioning. To update an existing event declaration, a new "Event Type" must be defined with a new "Event Descriptor". However, it is recommended that the updated "Event Type" maintain the same name followed by an incremented version number suffix. The description field of the new event type should be used to explain the difference between the two related “Event Descriptors”. 
+  4. The “Event Descriptor” must define a default time-to-live (TTL) for events of this type.
+  5. The “Event Descriptor” can optionally define data structure definitions or schemas for an event payload.
+  6. The “Event Descriptor” can optionally define events as “Signed” of “Unsigned” (not in scope of MVP, but the basis for that must be provided). “Signed” means that all events of this event type must be digitally signed by the publisher.
+
+- Each instantiated event must contain at least a UUID and an "Event Type". A payload for the event is optional and can be of any type of serializable data and it is up to a publisher and a consumer to agree on the data structures they want to exchange. If an event descriptor contains data structure definitions or schemas, they can be used for validation, marshalling/unmarshalling.
+
+- It should be possible to notify a publisher that there are no subscribers for an event type it is trying to send. It could be included in a response to an event publishing call.
+
+- Time to live (TTL) must be set up for each event sent. It optionally allows a publisher to provide a callback endpoint or an error “Event Type” to be notified that despite the fact that there are subscribers for such an event type no one has received the event within the specified period of time. If a TTL is not explicitly specified for an event, the default TTL for the event type will be used.
```

Comment:
This notifies the publisher that the published event is essentially dead. It can allow the publisher to trigger alternative logic, including in some cases a retry attempt with a longer TTL.

Reply:
Ok, upon this error, a publisher might choose to re-submit the same event, with a different UUID?

Is that why this notification only occurs when no subscribers have been delivered to, rather then some, because it could cause multiple deliveries of the same event?

Reply:
Yes

Reply:
Perhaps so it can re-publish the event (say up to N times)?